### PR TITLE
add a java mp jpa scenario using crunchy psql operator

### PR DESCRIPTION
**Pull request by ajm01**
fixes https://github.com/redhat-developer/service-binding-operator/issues/611

### Motivation

Introduce a new sample scenario - java openliberty microservice jpa app using a CrunchyData PostgreSQL Operator managed db and the Service Binding Operator. Also shows the use of odo to create the SBR, the intermediate secret object and how to configure the CruncyData PostgreSQL CR.

### Changes

added a readme document describing the process from start to finish.

### Testing

n/a at this stage

For further more details refer the CONTRIBUTING.md

### CODE REVIEW

File: examples/java_ol_postgresql_odo/README.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+odo push -f
+```
+
+Once the new version is up, display the openshift logs and notice a successful test run
+```shell
+odo log
+```
+Notice the tests are successful and contain data responses from the PostgreSQL database
+
+```shell
```
[the remark this answers is not in the document]
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: I have updated the sample to now have a UI, and documented it here int he README - it (the app) is not in our samples repo master branch but should be soon

File: examples/java_ol_postgresql_odo/README.md [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+> odo push
+```
+
+You have now created a Service Binding Request object called `jpa-servicebindingrequest-example-servicebindingrequest` in the cluster on behalf of your application. We must manually configure the YAML files associated with it and the Custom Resource Object associated with the database in order to link them both together.
+
+You can see this Service Binding Request via kubectl
+```shell
+> kubectl get servicebindingrequest jpa-servicebindingrequest-example-servicebindingrequest
+NAME                                                      AGE
+jpa-servicebindingrequest-example-servicebindingrequest   3m12s
```
[the remark this answers is not in the document]
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md:237]: added th eYAML for this below - can ship as a separate file if need be

**Comment by ajm01** [comment]
Hi @sbose78 - i think I have covered all comments, if you want to look this over again. I have the referenced app src code updated to include  a CRUD gui and it is merged into our sample repo now.

File: examples/java_ol_postgresql_odo/README.md [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+  generation: 2
+  namespace: readmeapp
+  ownerReferences:
+    - apiVersion: apps/v1
+      kind: Deployment
+      name: mysboproj
+      uid: e3ca87f2-0208-4b00-82bb-db4345f834dd
+  finalizers:
+    - finalizer.servicebindingrequest.openshift.io
+spec:
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:431]: There were some API changes that got introduced with https://github.com/redhat-developer/service-binding-operator/pull/608 . for eg, applicationselector to the application, backingServiceSelector to selectors. @ajm01 can you PTAL. :)

File: examples/java_ol_postgresql_odo/README.md [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+
+or running the following command to install the operator completely:
+
+``` shell
+make install-service-binding-operator-master
+```
+
+#### Install the DB operator
+
+Login to your Openshift terminal for all command line operations
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:80]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+
+Follow the installations instructions for [Installing Crunchy PostgreSQL for Kubernetes](https://operatorhub.io/operator/postgresql)
+
+When you have completed the `Before You Begin` instructions, you may access your Openshift Console and install the Crunchy PostgreSQL Operator from the Operator Hub:
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:84]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+
+![Service Binding Operator as shown in OperatorHub](../../assets/Crunchy.png)
+
+After the instalation completes via the Operator Hub, please follow the instructions in the `After You Install` section.
+
+### Application Developer
+
+#### Access your Openshift terminal and oc login to the Openshift Cluster
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:92]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+
+```shell
+pgo create cluster my-demo-db -n service-binding-demo
+```
+
+#### Import the demo Java MicroService JPA application
+
+In this example we will use odo to manage a sample [Java MicroServices JPA application](https://github.com/OpenLiberty/application-stack-samples.git).
+
+From the Openshift terminal, create a project directory `my-sample-jpa-app`
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:114]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+```
+cd to the sample JPA app
+```shell
+> cd ./application-stack-samples/java-openliberty/samples/jpa
+```
+initialize project using odo
+```shell
+> odo create java-openliberty mysboproj
+```
+Create an openshift route to access the application
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:133]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+```shell
+> odo url create --port 9080
+```
+
+Perform an initial odo push of the app to the cluster
+```shell
+> odo push 
+```
+
+The application is now deployed to the cluster - you can view the status of the cluster and the application test results by streaming the openshift logs to the terminal
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:143]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+[INFO] [err]    at com.ibm.websphere.jaxrs.server.IBMRestServlet.service(IBMRestServlet.java:96)
+[INFO] [err]    at [internal classes]
+[INFO] [err]    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
+[INFO] [err]    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
+[INFO] [err]    at java.base/java.lang.Thread.run(Thread.java:834)
+6590 INFO org.microshed.testing.jaxrs.JsonBProvider  - Response from server: []
+
+```
+
+You can also access the application via the openshift URL created ealier. To see the URL that was created, list it
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:170]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+You have now created a Service Binding Request object in the namespace on behalf of your application. We must manually configure the YAML files associated with it and the Custom Resource Object associated with the database in order to link them both together.
+
+You can see this Service Binding Request via kubectl
+```shell
+> kubectl get servicebindingrequest jpa-servicebindingrequest-example-servicebindingrequest
+NAME                                                      AGE
+jpa-servicebindingrequest-example-servicebindingrequest   3m12s
+>
+```
+Or, alternatively, you can inspect the SBR via the Openshift console in Administrator view by navigating to Operators > Installed Operators > Service Binding Operator and clicking on the Service Binding Request tab. Select the Service Binding Request Instance named `mysboproj-servicebindingrequest-example-servicebindingrequest`
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:240]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+
+#### Manually configure YAML files
+
+Access the Openshift console and navigate to Administration > Custom Resource Definitions > Pgcluster and click on the 'Instances' tab
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:244]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+
+You have now created an intermediate secret object called `mysboproj-servicebindingrequest-example-servicebindingrequest` in the cluster that can be used by your application. You can see this secret via kubectl
+
+```shell
+kubectl get secret mysboproj-servicebindingrequest-example-servicebindingrequest
+NAME                                                            TYPE     DATA   AGE
+mysboproj-servicebindingrequest-example-servicebindingrequest   Opaque   5      13m
+>
+```
+Or, alternatively, you can inspect the new intermediate secret via the Openshift console in Administrator view by navigating to Workloads > Secrets and clicking on the secret named `mysboproj-servicebindingrequest-example-servicebindingrequest` Notice it contains 5 pieces of data all related to the connection information for your PostgreSQL database instance.
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:467]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+
+Once the new version is up (there will be a slight delay until application is available), navigate to the CreatePerson.xhtml using the URL created in a previous step. Enter requested data and click the "Save" button
+![Create Person xhtml page](../../assets/createPersonDB.png)
+
+Notice you are re-directed to the PersonList.xhtml page, where your data is displayed having been input to the postgreSQL database and retrieved for display purposes.
+![Create Person xhtml page](../../assets/displayPeople.png)
+
+You may inspect the database instance itself and query the table to see the data in place by using the postgreSQL command line tool, psql.
+
+Navigate to the pod containing your db from the Openshift Console
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:482]: OpenShift

File: examples/java_ol_postgresql_odo/README.md [hunk 15]
```diff
@@ -0,0 +1,510 @@
...
+  ```
+Save this YAML file and reload it.
+
+Navigate to Operators > Installed Operators > Service Binding Operator and click on the 'Instances' tab
+
+Click on the Service Binding Request you created in the previous section and edit its YAML file
+
+We will edit the `backingServiceSelector` . This section is used to find the backing service - our operator-backed DB instance called `db-demo`.
+
+Replace the `backingServiceSelector` block with th efollowing YAML snippet:
```
Comment (Avni-Sharma) [inline examples/java_ol_postgresql_odo/README.md:402]: the following

File: examples/java_ol_postgresql_odo/README.md [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+```shell
+> odo catalog list services
+Operators available in the cluster
+NAME                                    CRDs
+service-binding-operator.v0.1.1-352     ServiceBindingRequest
+>
+```
+use odo to create an odo service for the Service Binding Operator
+```shell
+> odo service create service-binding-operator.v0.1.1-352/ServiceBindingRequest
```
[the remark this answers is not in the document]
Comment (dharmit) [reply examples/java_ol_postgresql_odo/README.md:207]: `odo link <service-name>` would link the component under `pwd` to the mentioned service. In this case, if you want to link the component `mysboproj` to, say, `EtcdCluster/myetcdcluster`, you would `cd` into `mysboproj`'s component directory and do:

```
$ odo link EtcdCluster/myetcdcluster
```

File: examples/java_ol_postgresql_odo/README.md (hunk 4 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 5 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 6 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 7 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 9 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 8 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 10 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 12 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 13 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 14 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 15 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 11 above)
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: fixed

File: examples/java_ol_postgresql_odo/README.md (hunk 3 above)
Comment (sbose78) [reply examples/java_ol_postgresql_odo/README.md:431]: @Avni-Sharma Would that be a validation of `master` or are the changes available on operatorhub ?

File: examples/java_ol_postgresql_odo/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply examples/java_ol_postgresql_odo/README.md:207]: Thanks,

@ajm01 , Here's how the sequence could look like

0. `odo service create <postgres>`
1.  update yaml with annotations
2. `odo link  Pgcluster/<name-of-resource`

You should not have to do `odo service create service-binding-operator.v0.1.1-352/ServiceBindingRequest`

[the remark this answers is not in the document]
Comment (ajm01) [reply examples/java_ol_postgresql_odo/README.md]: ok, i confirmed this is the correct sequence and have updated the doc to reflect that.

File: examples/java_ol_postgresql_odo/README.md (hunk 3 above)
Comment (Avni-Sharma) [reply examples/java_ol_postgresql_odo/README.md:408]: Validation of master and `beta` channel where we have the changes in `0.2.0-387` on operator-hub

Comment (Avni-Sharma) [reply examples/java_ol_postgresql_odo/README.md:408]: Also , all the examples/ in the repo have the new APIs now

File: examples/java_ol_postgresql_odo/README.md [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+```
+
+Save and re-load this YAML file.
+
+You have now created an intermediate secret object called `jpa-pgcluster-my-demo-db` in the cluster that can be used by your application. You can see this secret via kubectl
+
+```shell
+kubectl get secret jpa-pgcluster-my-demo-db
+NAME                                                            TYPE     DATA   AGE
+jpa-pgcluster-my-demo-db   Opaque   5      13m
```
Comment (sbose78) [inline examples/java_ol_postgresql_odo/README.md:441]: Could you print the contents of the secret please?

**Comment by pedjak** [comment]
@ajm01 can you rework this PR to match the current API please?

**Comment by Avni-Sharma** [comment]
@pedjak I think now, we need a rework for the API and annotations as well

**Comment by ajm01** [comment]
closing this PR for now due to its use of odo and the downlevel SBO. Will work to understand if the sample can be re-written to avoid the use of odo and make use of the latest updates to SBO and its annotation updates.

**Pull request closed without merging**